backend/src/main/java/fk/prof/backend/model/request/RecordedProfileHeaderParser.java

```diff
@@ -47,60 +47,77 @@ public RecordedProfileHeader get() {
           if (headerLength < 1 || headerLength > maxAllowedBytesForRecordingHeader) {
-            throw new HttpFailure("Allowed range for recording header length is 1B to " + maxAllowedBytesForRecordingHeader + "B", 400);
+            throw new AggregationFailure("Allowed range for recording header length is 1B to " + maxAllowedBytesForRecordingHeader + "B");
           }
-          currentPos = updateChecksumAndGetNewPos(underlyingBuffer, codedInputStream, currentPos);
+          byte[] headerLengthBytes = inputStream.getBytesReadSinceMark();
+          headerChecksum.update(headerLengthBytes, 0, headerLengthBytes.length);
         }
 
         if (recordingHeader == null) {
-          if (headerLength > (underlyingBuffer.length() - currentPos)) {
-            return currentPos;
+          if(inputStream.available() < headerLength) {
+            return;
           }
-
-          int oldLimit = codedInputStream.pushLimit(headerLength);
           try {
-            recordingHeader = Recorder.RecordingHeader.parseFrom(codedInputStream);
-          } catch (IOException ex) {
+            inputStream.mark(0);
+            byte[] headerBytes = new byte[headerLength];
+            inputStream.read(headerBytes, 0, headerLength);
+            recordingHeader = Recorder.RecordingHeader.parseFrom(headerBytes);
+            headerChecksum.update(headerBytes, 0, headerBytes.length);
+          } catch (InvalidProtocolBufferException ex) {
             //Running buffer has sufficient bytes present for reading recording header. If exception is thrown while parsing, send error response
-            throw new HttpFailure("Error while parsing recording header", 400);
+            throw new AggregationFailure("Error while parsing recording header");
           }
-          codedInputStream.popLimit(oldLimit);
-          currentPos = updateChecksumAndGetNewPos(underlyingBuffer, codedInputStream, currentPos);
         }
 
         if (checksumValue == null) {
-          checksumValue = codedInputStream.readUInt32();
-          currentPos = codedInputStream.getTotalBytesRead();
+          inputStream.mark(0);
+          int firstByte = inputStream.read();
+          if(firstByte == -1) {
+            throw new InvalidProtocolBufferException("EOF when reading RecordingHeader:checksum from inputstream");
+          }
+          checksumValue = CodedInputStream.readRawVarint32(firstByte, inputStream);
           if ((int) headerChecksum.getValue() != checksumValue) {
-            throw new HttpFailure("Checksum of header does not match", 400);
+            throw new AggregationFailure("Checksum of header does not match");
           }
           headerParsed = true;
         }
       }
     } catch (IOException ex) {
-      //NOTE: Ignore this exception. Can come because incomplete request has been received. Chunks can be received later
+      if(!(ex instanceof InvalidProtocolBufferException)) {
```

Comment:
why not do the following:
try {...
} catch (InvalidProtocolBufferException e) {
try {inputStream.reset();} catch(IOException e1) {}
} catch (IOException e) {
  throw new AggregationFailure(e);
}

This seems a lot simpler to read and understand.